### ISSUE TITLE
`while-let` false positives for irrefutable patterns

### ISSUE THREAD

**Issue by Zoxc** [Manishearth/rust-clippy#800]
Loops with irrefutable let bindings are suggested to be changed into `while let`, which is not possible.

For example this loop:
```rust
loop {
    let (attr, form) = match try!(read_abbrev_pair(adata)) {
        Some(p) => p,
        None => break
    };
}
```

**Comment by mcarton** [comment]
In that particular case you can write `while let Some((attr, form)) = try!(read_abbrev_pair(adata))`.

We already have #771 suggesting that lint is too lax, so I’m closing this.

**Comment by Zoxc** [comment]
The suggestion given is at least wrong: `while let Some(p) = try!(read_abbrev_pair(adata)) { .. }`
